presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
-            TableWriteInfo tableWriteInfo,
-            Map<PlanFragmentId, Broadcast<List<PrestoSparkSerializedPage>>> broadcastInputs)
+            PartitioningHandle partitioning,
+            List<TableScanNode> tableScans,
+            Optional<Integer> expectedNumberOfPartitions)
     {
-        checkInputs(fragment.getRemoteSourceNodes(), ImmutableMap.of(), broadcastInputs);
-
-        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
-        checkArgument(
-                tableScans.size() == 1,
-                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
-                fragment.getId(),
-                tableScans.size());
-
-        TableScanNode tableScan = getOnlyElement(tableScans);
-
-        List<ScheduledSplit> splits = getSplits(session, tableScan);
-        shuffle(splits);
-        int initialPartitionCount = getSparkInitialPartitionCount(session);
-        int numTasks = Math.min(splits.size(), initialPartitionCount);
-        if (numTasks == 0) {
-            return JavaPairRDD.fromJavaRDD(sparkContext.emptyRDD());
+        ListMultimap<Integer, TaskSource> taskSourcesMap = ArrayListMultimap.create();
+        for (TableScanNode tableScan : tableScans) {
+            List<ScheduledSplit> scheduledSplits = getSplits(session, tableScan);
+            shuffle(scheduledSplits);
+            SetMultimap<Integer, ScheduledSplit> assignedSplits = assignSplitsToTasks(session, partitioning, scheduledSplits);
+            asMap(assignedSplits).forEach((partitionId, splits) ->
+                    taskSourcesMap.put(partitionId, new TaskSource(tableScan.getId(), splits, true)));
         }
 
-        List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(splits, numTasks);
-
-        // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
-        splits = null;
-
-        ImmutableList.Builder<SerializedPrestoSparkTaskDescriptor> serializedTaskDescriptors = ImmutableList.builder();
-        for (int i = 0; i < assignedSplits.size(); i++) {
-            List<ScheduledSplit> splitBatch = assignedSplits.get(i);
-            PrestoSparkTaskDescriptor taskDescriptor = createSourceTaskDescriptor(session, tableWriteInfo, fragment, splitBatch);
-            // TODO: consider more efficient serialization or apply compression to save precious memory on the Driver
-            byte[] jsonSerializedTaskDescriptor = taskDescriptorJsonCodec.toJsonBytes(taskDescriptor);
-            serializedTaskDescriptors.add(new SerializedPrestoSparkTaskDescriptor(jsonSerializedTaskDescriptor));
-            // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
-            assignedSplits.set(i, null);
+        IntStream partitions = expectedNumberOfPartitions
```

Comment:
As you might guess, I will use more "plain" style `if`,`list`,`for-loop` for line 290-303 😃 . But it's personal taste.

Reply:
Also, would it help understand code if bucketed table / unbucketed table split assignment is done in different code blocks? -- while they share the concept of "partitions", its semantic is different. For bucketed tables, you have to have the same number of partitions as with buckets, while for unbucketed tables, the number of partition is by configured or automatically tuned (e.g. based on table size).

Reply:
I actually wanted to use `IntStream` specifically to avoid materializing `IntStream.range(0, integer)`

Reply:
> I actually wanted to use IntStream specifically to avoid materializing IntStream.range(0, integer)

 Why not just using for loop? ...

Reply:
You will have to loop either over the partitions from `taskSourcesMap` or over all partitions provided by `expectedNumberOfPartitions `. With simple `for` loop it seems like you would have to duplicate the loop body twice.

Reply:
@arhimondr : Stream API often yields non-trivial performance overhead (see, for example https://github.com/prestodb/presto/pull/11374) . And I don't think this part of code is performance sensitive? (it's just doing plan once on Driver?)

I personally prefer not using Stream API unless it's pretty clear (e.g. the classic `.map().filter().XXX` pattern. This has also been discussed in https://github.com/prestodb/presto/pull/11374#issuecomment-418464932 . And I also have expressed similar ideas in my other code reviews. 

I understand a general policy discussion about Stream API should be done by TSC. 

In this specific case, I found it make hard for me to follow the core logic with Stream API 😕

Reply:
@arhimondr : Java Stream API can also have non-trivial memory allocation overhead, see https://github.com/prestodb/presto/pull/13984

Reply:
Replaces with 2 branches and regular loops. Added comments.